### DISCUSSION TITLE
Is it possible to explicitly specify Erlang version (ERLANG_HOME?) for RabbitMQ .exe installer on Windows?

### DISCUSSION THREAD

**Discussion by inikulshin** [rabbitmq/rabbitmq-server#2753] [Questions] [answered]
~~[Make Sure ERLANG_HOME is Set](https://www.rabbitmq.com/install-windows-manual.html#set-erlang-home-variable) didn't work for me.~~

@lukebakken 
> The instructions here specify that you are using the `.zip` file installer (link) and NOT the `.exe` installer.

However this difference between `.zip` and `.exe` installers is pretty unexpected.

It looks like `ERLANG_HOME` environment variable is ignored and the chosen Erlang version depends only on order of **EnumRegKey** in
[packaging/windows-exe/rabbitmq_nsi.in](https://github.com/rabbitmq/rabbitmq-server-release/blob/master/packaging/windows-exe/rabbitmq_nsi.in)

```
Function findErlang

  StrCpy $0 0
  StrCpy $2 "not-found"
  ${Do}
    EnumRegKey $1 HKLM Software\Ericsson\Erlang $0
    ${If} $1 = ""
      ${Break}
    ${EndIf}
    ${If} $1 <> "ErlSrv"
      StrCpy $2 $1
    ${EndIf}

    IntOp $0 $0 + 1
  ${Loop}

```

And ERLANG_HOME is just overridden in the end:

`WriteRegExpandStr ${env_hklm} ERLANG_HOME $0`

**Comment by lukebakken** [comment]
The instructions [here](https://www.rabbitmq.com/install-windows-manual.html#set-erlang-home-variable) specify that you are using the `.zip` file installer ([link](https://www.rabbitmq.com/install-windows-manual.html#install-rabbitmq)) and NOT the `.exe` installer.

**Comment by inikulshin** [comment]
@lukebakken you're right, the first line in not related to `.exe` installer, fixed.

But the rest is. And this difference between `.zip` and `.exe` installers is pretty unexpected.
I've updated title question.

**Comment by lukebakken** [comment, marked as answer]
As you pointed out, the installer will set the ERLANG_HOME environment variable. The workaround is to install RabbitMQ yourself using the zip file and associated rabbitmq-service.bat file to install the Windows service.

**Reply by inikulshin** [reply]
It's completely ok, that `.exe` installer sets ERLANG_HOME environment variable, when **it is not set**. (However, the choice from multiple Erlangs could be more wise than "the one with the greatest index in EnumRegKey". For example, the one with the maximum supported version.)

But why `.exe` installer can't just keep existing value of ERLANG_HOME? Don't you think it is a bug?

**Reply by lukebakken** [reply]
> But why .exe installer can't just keep existing value of ERLANG_HOME? Don't you think it is a bug?

You are the first person to point this out. It's not a bug, just the way the installer has always worked. Using the `zip` file must be sufficient for most users who wish to keep their existing `ERLANG_HOME`.

**Reply by inikulshin** [reply]
I suppose, multiple Erlang installations are very rare. Most of RabbitMQ users don't need more than one, and never had.
We also discovered this behavior by accident mistake, and it doesn't really blocks us.

And it is not a bug:

https://www.rabbitmq.com/install-windows.html#installer

> * There must be only one Erlang version installed at a time

Thank you.

**Reply by inikulshin** [reply]
Actually, .exe installer iterates only on 64-bit Erlangs, so the prerequisite is 

> There must be only one **64-bit** Erlang version installed at a time

**Reply by michaelklishin** [reply]
No one should ever run 32-bit Erlang for RabbitMQ (or anything else). 64-bit is the implicit assumption and it has been for years. The limited memory address space makes 32-bit a no-go for data services.

**Reply by inikulshin** [reply]
@michaelklishin sure.

I just wanted to mention, that any number of installed 32-bit Erlangs (in addition to the only one prerequisite 64-bit) does not affect `.exe` installer.

Erlang is a different product, people may install and use it regardless RabbitMQ...

**Comment by lukebakken** [comment]
If you'd like to open a pull request to make the installer detect the presence of `ERLANG_HOME` I think that would be reasonable (cc @michaelklishin).
